### ISSUE TITLE
Speed limits

### ISSUE THREAD

**Issue by Starcommander** [junjunguo/PocketMaps#144]
As suggested in #142 it is an idea to support speed limits.
But this is not an easy task.

Some considerations:
https://wiki.openstreetmap.org/wiki/Key:maxspeed
https://wiki.openstreetmap.org/wiki/Default_speed_limits

**Comment by Starcommander** [comment]
Now it is possible to get details of path:
`Navigator.getNavigator().getGhResponse().getPathDetails()`
This will return the following entries:
`max_speed=[null [0, 38], 100.0 [38, 62], 70.0 [62, 66], null [66, 71], 50.0 [71, 94]]`
`average_speed=[30.0 [0, 2], 50.0 [2, 37], 65.0 [37, 38], 90.0 [38, 62], 65.0 [62, 66], 50.0 [66, 71], 45.0 [71, 94]]`
Seems, that **average_speed** is always present, where **max_speed** is sometimes missing.

As a first step I would suggest to just show max_speed where present, and average_speed with quest-mark, when missing.

Open questions
- Is it possible to implement the rules from above links for missing speed-limits?
- Switching from meters to miles, is the result still km/h also in different countries?

**Comment by Starcommander** [comment]
Now speed-sign is showning in newest pre-release.
The apk is availhable here:
https://github.com/junjunguo/PocketMaps/releases/tag/3.6-pre1

@stanrobertdev Do you want to test the new feature?
@feelingwalnut As you are using miles instead of meters, I would be very interested, what is showing on the speed-sign, and whether this is correct.

**Comment by feelingwalnut** [comment]
In imperial the speed is still in km/h instead of MPH. 
it appears "1 kilometer per hour = 0.62137 miles per hour"
the speeds look accurate once the math is done, many small roads show a speed with a number and a ?, but they also look correct.
the speed sign itself is a red circle, that is not what we use in America but it is clear what the purpose is.

voice does indicate "feet" properly now.

for me debugsimulator simulates all routes, not just the first. a debug version with simulator off or toggle would be good for real life testing.

when you reach a destination the screen sits there at "navigation end" before this, the voice will prompt you that navigation is ending in "X distance" but there is no confirmation your nav has ended or a return to the main map. **is this a bug?**

when you close the app and reopen, it goes back to metric from imperial. **settings not saved?**

**Comment by Starcommander** [comment]
@feelingwalnut Thank you for your testing, and sharing the results!
So I have to mult the value with 0.62137 before showing the value, right?
Can you give me some example values that are shown?
The question-mark '?' indicates, that average-speed is shown instead of max-speed.

The debug-simulator simulates the first route every time.
So if you create a different route, the debug-simulator moves along the first route again.
This is for debugging what happens, when the car moves a wrong way.
Of course the route will recalculate, in this case.
Currently you have to kill the app, for simulating a different route.
Options for debug-simulator: A good idea for later.

Basically all instructions (turn_left, turn_right, navigation_end) are only spoken before the "X-distance" so I do not bother the user with repeating to speak same instruction.

I was not able to reproduce the issue, where settings are not saved.
Can you explain exactly, how you "close the app and reopen" ?
Thanks for your help.

**Comment by feelingwalnut** [comment]
I'm not certain what the formula is for mph to km/h
https://duckduckgo.com/?t=ffab&q=mph+to+kmh&ia=answer
this shows 1.609344. divide or multiply by the correct number to convert from km/h to mph.
last test it was displaying 70km/h in an area that is 20mph, that is pretty inaccurate, it may have been an isolated issue with that short road.
on the highway it displayed 80km/h, the speed limit there is 50mph which would make the 1.609344 number correct. on another highway it displayed 115 km/h for a speed of 71.46mph. speed limit on that highway is 70mph.
American limits are typically in increments of 5mph. 5, 10, 15, 20, 25, up to about 85mph. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States
so you may have to take the speeds provided, do the conversion and then round to the nearest increment of 5 to avoid having decimals and strange numbers. (sorry, probably a lot of work)

the virtual car did go the wrong way several times as i was playing around. the app recalculated properly except one instance the car just drove where it wanted and ignored all the recalculations so the app spent a lot of time telling the car to turn around.

when i close an app i'll pull up the running apps with the three white bars button and swipe the app away, such as method 1 in this link.
https://www.wikihow.tech/Close-Apps-on-Android
on android 10, samsung galaxy s9 (not root)

amusing fact about speed limit is that in my area we don't listen to speed limits. 
we drive how we please, sometimes double or even triple the limit. many parts of America are like this in my experience.

the vocal instructions including navigation_end are fine. confirmed it was not a bug.

**Comment by Starcommander** [comment]
@feelingwalnut 
Converting and rounding the value is really not such a big task. --> Done.
Seems that the stored values, that I get from the map are all in km/h rounded to mult of 5?!

Recalculating the route is the expected behaviour regarding DebugSimulator.
The car - of course - will ignore any new route and drives the first route like a dummy. ;)

Basically the unit (imperal/metric) is directly stored in settings after selection.
So it should be stored correctly in any way.

Maybe you have two pocketmaps-versions installed on your phone?
This is possible with pocketmaps from google-play-store because of different app name.
It is the only cause, I can think of, currently.

**Comment by feelingwalnut** [comment]
i did have f-droid version installed before. maybe there is a back end conflict on my phone with tidbits of files and folders. i will look into diagnosing it.

the dummy car is amusing.

**Comment by Starcommander** [comment]
Ok. If you can find out more, let me know.
Thanks.

**Comment by feelingwalnut** [comment]
should probably make a new issue but i found the "not saving metric/imperial" option only applies to 3.6pre1. 
tested 3.5pre1 and 3.4 release, they functioned normal.
could not find 3.5 release on fdroid or google play. did not try to build it from source.
could not find regression.
replicated on two different device, physical phone and bluestacks emulator.

**Comment by Starcommander** [comment]
@feelingwalnut Thanks for your effort.
I will re-check all changes between 3.5pre1 and 3.6pre1
Unfortunately 3.5 is still not released on FDroid and there is still no progress.
I will open an issue on their page.

**Comment by Starcommander** [comment]
I have checked the differences of both versions.
Most likely I suspect the voice-settings, they may lead to this issue.
So there are two ways for debugging:
- Creating a logcat output directly after changing the unit from metric to imperal. Maybe this will print an exception directly.
- Select a voice and a tts-engine explicitly in main settings, and trying again to change unit. Maybe the issue does not occur afterwards.

@feelingwalnut 
Are you able to provide such test-results?
Both tests should be executed in version 3.6pre1
This would be very helpful.

**Comment by feelingwalnut** [comment]
choosing voice in menu does still cause issue, do you want logcat dumped here or uploaded somewhere?
i used adb logcat | findstr com.junjunguo.pocketmaps

**Comment by Starcommander** [comment]
@feelingwalnut Logcat output may help to find the issue.
Dump it here, or attach it as file.
Thanks in advance.

**Comment by feelingwalnut** [comment]
[logcat.txt](https://github.com/junjunguo/PocketMaps/files/4638223/logcat.txt)
try that.

**Comment by Starcommander** [comment]
@feelingwalnut  Thanks!
As I thought, it has to do with tts-voice-settings.
On your logcat output the exception occurs in line 69:
- Error at Variable.java:565

In source-code this line 565 is:
- setTtsEngine(jo.getString("ttsEngine"));

So maybe the tts-engine is still not saved in json-string.
I will solve this in next release.
Thanks again for your help.

**Comment by Starcommander** [comment]
New release 3.7 is available in F-Droid, and soon on Play-Store.
